crates/accelerate/src/circuit_library/n_local/blocks.rs

```diff
@@ -0,0 +1,171 @@
+    ) -> PyResult<(PackedOperation, SmallVec<[Param; 3]>)> {
+        match self {
+            Self::Standard { gate } => Ok((
+                (*gate).into(),
+                SmallVec::from_iter(params.iter().map(|&p| p.clone())),
+            )),
+            Self::PyCustom { builder } => {
+                // the builder returns a Python operation plus the bound parameters
+                let py_params =
+                    PyList::new_bound(py, params.iter().map(|&p| p.clone().into_py(py))).into_any();
+
+                let job = builder.call1(py, (py_params,))?;
+                let result = job.downcast_bound::<PyTuple>(py)?;
+
+                let operation: OperationFromPython = result.get_item(0)?.extract()?;
+                let bound_params = result
+                    .get_item(1)?
+                    .iter()?
+                    .map(|ob| Param::extract_no_coerce(&ob?))
+                    .collect::<PyResult<Vec<Param>>>()?;
+
+                Ok((
+                    operation.operation,
+                    SmallVec::<[Param; 3]>::from_vec(bound_params),
+                ))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+#[pyclass]
+pub struct Block {
+    pub operation: BlockOperation,
+    pub num_qubits: u32,
+    pub num_parameters: usize,
+}
+
+#[pymethods]
+impl Block {
+    #[staticmethod]
+    #[pyo3(signature = (gate,))]
+    pub fn from_standard_gate(gate: StandardGate) -> Self {
+        Block {
+            operation: BlockOperation::Standard { gate },
+            num_qubits: gate.num_qubits(),
+            num_parameters: gate.num_params() as usize,
+        }
+    }
+
+    #[staticmethod]
+    #[pyo3(signature = (num_qubits, num_parameters, builder,))]
+    pub fn from_callable(
+        py: Python,
+        num_qubits: i64,
+        num_parameters: i64,
+        builder: &Bound<PyAny>,
+    ) -> PyResult<Self> {
+        if !builder.is_callable() {
+            return Err(QiskitError::new_err(
+                "builder must be a callable: parameters->(bound gate, bound gate params)",
+            ));
+        }
+        let block = Block {
+            operation: BlockOperation::PyCustom {
+                builder: builder.to_object(py),
+            },
+            num_qubits: num_qubits as u32,
+            num_parameters: num_parameters as usize,
+        };
+
+        Ok(block)
+    }
+}
+
+type BlockEntanglement = Vec<Vec<u32>>; // vector of connections for a block (e.g. CX)
+pub(super) type LayerEntanglement = Vec<BlockEntanglement>; // entanglement of all blocks
+
+/// Represent the entanglement in an n-local circuit.
+///
+/// This eagerly expands the full entanglement for every layer.
+/// This could be done more efficiently, e.g., by creating entanglement objects that store
+/// their underlying representation (e.g. a string or a list of connections) and returning
+/// these when given a layer-index.
+pub struct Entanglement {
+    entanglement_vec: Vec<LayerEntanglement>,
```

Comment:
Correct:
```
Connection: Vec<u32>  // indices that the multi-qubit gate acts on
BlockEntanglement: Vec<Connection>  // entanglement for single block
LayerEntanglement: Vec<BlockEntanglement>  // entanglements for all blocks in the layer
Entanglement: Vec<LayerEntanglement>  // entanglement for every layer
```
Maybe this would be nice to add as explanatory comment?

Reply:
A comment wouldn't hurt. I assume you're using the type aliases here to make it simpler. It would be easier for me personally to understand of `Vec<u32>` `Vec<Vec<u32>>`, etc rather than `Connection`, `BlockEntanglement`, etc. But I guess cargo doesn't like that. Either way if you want to add a comment that would be nice for people reading the code.

Reply:
yeah I also found myself going back checking whether `BlockEntanglement` is now a 2 or 3 times nested Vec, but I think in the end the explanation plus the types might make it a bit easier to read 🙂